src/sys/autokill.zig

```diff
@@ -0,0 +1,248 @@
+const KillPass = enum {
+    sigterm, // Send SIGTERM for graceful shutdown
+    sigstop, // Send SIGSTOP to freeze processes
+    sigkill, // Send SIGKILL for forced termination
+};
+
+/// Kill all child processes using a three-pass strategy:
+///
+/// 1. SIGTERM: Graceful shutdown - allows cleanup handlers to run (500μs delay)
+/// 2. SIGSTOP: Freeze survivors - prevents reparenting races
+/// 3. SIGKILL: Force termination - ensures nothing survives
+///
+/// Each pass freshly enumerates children to catch any spawned during the sequence.
+/// Early bailout if no children remain after any pass.
+///
+/// This is more graceful than SIGSTOP→SIGKILL (allows cleanup) and more thorough
+/// than SIGTERM→SIGKILL (SIGSTOP prevents races). Most processes exit from SIGTERM,
+/// making this faster in practice despite being three passes.
+pub fn killAllChildProcesses() void {
+    if (Environment.isWindows) {
+        // Windows already uses Job Objects which automatically kill children on exit
+        // This is a no-op
+        return;
+    }
+
+    const current_pid = std.c.getpid();
+
+    // Walk the process tree and kill only child processes
+    // Do NOT kill the entire process group with kill(-pid) as that would
+    // kill the Bun process itself before it can finish shutting down
+
+    // Pass 1: SIGTERM to allow graceful cleanup
+    // Give processes a chance to handle cleanup work before forced termination
+    {
+        const children = getChildPids(current_pid, current_pid) catch &[_]c_int{};
+        defer if (children.len > 0) bun.default_allocator.free(children);
+
+        if (children.len > 0) {
+            var seen = std.AutoHashMap(c_int, void).init(bun.default_allocator);
+            defer seen.deinit();
+            for (children) |child| {
+                killProcessTreeRecursive(child, &seen, current_pid, .sigterm) catch {};
+            }
+        }
+    }
+
+    // Brief delay to allow processes to handle SIGTERM
+    // Use longer delay on musl due to slower syscalls and /proc inconsistencies
+    const delay_us = if (Environment.isMusl) 2000 else 500;
+    std.time.sleep(delay_us * std.time.ns_per_us);
+
+    // Pass 2: SIGSTOP to freeze entire tree and minimize reparenting races
+    // Get fresh child list in case some exited from SIGTERM
+    {
+        const children = getChildPids(current_pid, current_pid) catch &[_]c_int{};
+        defer if (children.len > 0) bun.default_allocator.free(children);
+
+        if (children.len > 0) {
+            var seen = std.AutoHashMap(c_int, void).init(bun.default_allocator);
+            defer seen.deinit();
+            for (children) |child| {
+                killProcessTreeRecursive(child, &seen, current_pid, .sigstop) catch {};
+            }
+        }
+    }
+
+    // Pass 3: SIGKILL to force termination of any remaining processes
```

Comment:
With an unprivileged user id a process can be launched using sudo to change user id - then `SIGKILL` will fail with `EPERM`

Is there a way this condition can be reported back to the user?